### DISCUSSION TITLE
Github Actions Safety?

### DISCUSSION THREAD

**Discussion by bradlitterell** [tpm-rs/tpm-rs#186] [Organization]
See this guidance about preventing getting pwnd by 3rd party public PRs... because, well, that would be sad.

https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

A suggestion that was made to me privately was to limit the CI actions to being run by an pre-set group of owners to ensure that the CI pipeline only runs on PRs we have reviewed.  

Does anyone have any experience here with the best practice for this? 
In particular, please review the CI definition in https://github.com/tpm-rs/tpm-rs/pull/8.

I think it would be good to have additional eyeballs on this who are more experienced with github action safety than I.

**Comment by vsukhoml** [comment]
I found this link:
https://docs.github.com/en/enterprise-cloud@latest/admin/policies/enforcing-policies-for-your-enterprise/enforcing-policies-for-github-actions-in-your-enterprise

Some other ideas to consider:
[Enable Branch Protection](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/managing-a-branch-protection-rule) for default and release branches on your repo to require multi-party review for merges. Use the following settings:

- Require a pull request before merging
- Require approvals (select at least 1-2 as the required number of approvals before merging).
- Require status checks to pass before merging.
- Select “Include administrators” to apply the rules to people with admin permissions to the repository (the default excludes administrators from branch protection requirements).
- Require branches to be up to date before merging.
- Do NOT enable the “allow force pushes” or “allow deletions” settings (keep the default unchecked settings).

Not sure if helpful for Rust:
[Install the Scorecards Action](https://github.com/ossf/scorecard-action#installation) on your public repositories and [investigate the results](https://github.com/ossf/scorecard-action#view-results). [Scorecard](https://securityscorecards.dev/) is an automated tool that scans your project for security best practices. The GitHub Action runs a Scorecard scan on each change to the repository so you can monitor whether code changes introduce new security issues. 

Enable [Dependabot](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/about-dependabot-version-updates) on all org repositories.
First [create a dependabot.yml config file](https://docs.github.com/en/enterprise-server@3.3/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/enabling-and-disabling-dependabot-version-updates#enabling-dependabot-version-updates) to allow updates for dependencies. (You can use [this template](https://github.com/ossf/scorecard/blob/main/.github/dependabot.yml) for the config file.) Place the config file under the .github/ folder.
Then [enable the settings](https://docs.github.com/en/organizations/keeping-your-organization-secure/managing-security-settings-for-your-organization/managing-security-and-analysis-settings-for-your-organization), selecting options to enable Dependabot alerts and security updates. Also enable secret scanning if available.

[Enable CodeQL](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/setting-up-code-scanning-for-a-repository#setting-up-code-scanning-using-actions) or a similar, ecosystem-specific vulnerability scanning tool.

**Comment by vsukhoml** [comment]
https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions
